cpp/src/copying/contiguous_split.cu

```diff
@@ -894,20 +895,16 @@ std::vector<contiguous_split_result> contiguous_split(cudf::table_view const& in
   size_t total_temp_size = indices_size + src_buf_info_size + buf_sizes_size + dst_buf_info_size +
                            src_bufs_size + dst_bufs_size;
 
+  // clang-format off
   // allocate host
   std::vector<uint8_t> host_buf(total_temp_size);
-  // distribute
+  // distribute  
   uint8_t* cur_h_buf   = host_buf.data();
-  size_type* h_indices = reinterpret_cast<size_type*>(cur_h_buf);
-  cur_h_buf += indices_size;
-  src_buf_info* h_src_buf_info = reinterpret_cast<src_buf_info*>(cur_h_buf);
-  cur_h_buf += src_buf_info_size;
-  size_t* h_buf_sizes = reinterpret_cast<size_t*>(cur_h_buf);
-  cur_h_buf += buf_sizes_size;
-  dst_buf_info* h_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_h_buf);
-  cur_h_buf += dst_buf_info_size;
-  uint8_t** h_src_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);
-  cur_h_buf += src_bufs_size;
+  size_type* h_indices = reinterpret_cast<size_type*>(cur_h_buf);             cur_h_buf += indices_size;
+  src_buf_info* h_src_buf_info = reinterpret_cast<src_buf_info*>(cur_h_buf);  cur_h_buf += src_buf_info_size;
+  size_t* h_buf_sizes = reinterpret_cast<size_t*>(cur_h_buf);                 cur_h_buf += buf_sizes_size;
+  dst_buf_info* h_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_h_buf);  cur_h_buf += dst_buf_info_size;
+  uint8_t** h_src_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);              cur_h_buf += src_bufs_size;
```

Comment:
How about something like this: https://godbolt.org/z/he3TW7
```suggestion
   pack(host_buf.data(), h_indices, h_src_buf_info, h_buf_sizes, h_dst_buf_info, h_src_bufs);
```

Reply:
The `swallow` pattern isn't obvious the first time you see it. Louis Dionne does a good job of explaining it here: https://codereview.stackexchange.com/questions/51407/stdtuple-foreach-implementation

Reply:
Looks like I need to do a little modification to allow the size of each increment to be specified. Will update.

Reply:
Okay, here's what I came up with: https://godbolt.org/z/Pqdnj1

This can be used to replace the logic of calculating the size of the buffer in addition to doing the aliasing.

Reply:
I'll make the changes, but philosophically here's why I think it's the wrong thing to do:

- This is an absolutely trivial piece of code.  Yes, it's abstractable but this is really really granular.
- This will interrupt the flow of reading the code.  Instead of seeing :  set sizes, set pointers, set pointers, all using trivially readable arithmetic.   With this new method, now you have to go hunt down some external class to figure out what it's doing.  And the call itself, to me is unnecessarily mysterious:

```
    auto size = alias_pointers(nullptr, 4, {{&i, sizeof(int)}, 
                                            {&f, sizeof(float)}, 
                                            {&d, sizeof(double)}});
```
I _have_ to stop what I'm doing, set a bookmark, go find that function, read the docs, etc.   With the original code you immediately see "oh, this is just incrementing some pointers".

Reply:
> This is an absolutely trivial piece of code.

Just because it is doing simple things does _not_ make it trivial. Here's the original block of code with a minor change:
```
  // compute # of source buffers (column data, validity, children), # of partitions
  // and total # of buffers
  size_type num_src_bufs = count_src_bufs(input.begin(), input.end());
  size_t num_bufs        = num_src_bufs * num_partitions;

  // compute total size of host-side temp data
  size_t indices_size =
    cudf::util::round_up_safe((num_partitions + 1) * sizeof(size_type), split_align);
  size_t src_buf_info_size =
    cudf::util::round_up_safe(num_src_bufs * sizeof(src_buf_info), split_align);
  size_t buf_sizes_size = cudf::util::round_up_safe(num_partitions * sizeof(size_t), split_align);
  size_t dst_buf_info_size =
    cudf::util::round_up_safe(num_bufs * sizeof(dst_buf_info), split_align);
  size_t src_bufs_size = cudf::util::round_up_safe(num_src_bufs * sizeof(uint8_t*), split_align);
  size_t dst_bufs_size = cudf::util::round_up_safe(num_partitions * sizeof(uint8_t*), split_align);
  size_t total_temp_size = indices_size + src_buf_info_size + buf_sizes_size + dst_buf_info_size +
                           src_bufs_size;

  // clang-format off
  // allocate host
  std::vector<uint8_t> host_buf(total_temp_size);
  // distribute  
  uint8_t* cur_h_buf   = host_buf.data();
  size_type* h_indices = reinterpret_cast<size_type*>(cur_h_buf);             cur_h_buf += indices_size;
  src_buf_info* h_src_buf_info = reinterpret_cast<src_buf_info*>(cur_h_buf);  cur_h_buf += src_buf_info_size;
  size_t* h_buf_sizes = reinterpret_cast<size_t*>(cur_h_buf);                 cur_h_buf += buf_sizes_size;
  dst_buf_info* h_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_h_buf);  cur_h_buf += dst_buf_info_size;
  uint8_t** h_src_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);              cur_h_buf += src_bufs_size;
  uint8_t** h_dst_bufs = reinterpret_cast<uint8_t**>(cur_h_buf);

  // compute stack space needed for nested list offset calculation (needed on gpu only)
  int offset_stack_partition_size = compute_offset_stack_size(input.begin(), input.end());
  int offset_stack_size           = offset_stack_partition_size * num_partitions * 4;

  // allocate device
  size_t total_device_temp_size = total_temp_size + offset_stack_size;
  rmm::device_buffer device_buf{total_device_temp_size, stream, mr};
  // distribute
  uint8_t* cur_d_buf   = reinterpret_cast<uint8_t*>(device_buf.data());
  size_type* d_indices = reinterpret_cast<size_type*>(cur_d_buf);             cur_d_buf += indices_size;
  src_buf_info* d_src_buf_info = reinterpret_cast<src_buf_info*>(cur_d_buf);  cur_d_buf += src_buf_info_size;
  size_t* d_buf_sizes = reinterpret_cast<size_t*>(cur_d_buf);                 cur_d_buf += buf_sizes_size;
  dst_buf_info* d_dst_buf_info = reinterpret_cast<dst_buf_info*>(cur_d_buf);  cur_d_buf += dst_buf_info_size;
  uint8_t** d_src_bufs = reinterpret_cast<uint8_t**>(cur_d_buf);              cur_d_buf += dst_bufs_size;
  uint8_t** d_dst_bufs = reinterpret_cast<uint8_t**>(cur_d_buf);              cur_d_buf += dst_bufs_size;
  size_type* d_offset_stack = reinterpret_cast<size_type*>(cur_d_buf);
  // clang-format on
```

How long did it take for you to notice what was wrong? 

Anyone who looks at this is just going to see a wall of text that their eyes are going to glaze over and miss obvious errors. 

The original code is like a house built with duct tape and popsicle sticks. It's built out of simple things that anyone can look at and understand, but that doesn't make it a good house. 

The hallmark of good design is coming up with abstractions that make things easy to use and difficult to misuse. The original code is neither either to use (requires an incredible amount of manual effort) and very easy to misuse (see above comments about spotting errors).

Reply:
>  Code atomization like this works against comprehensibility when done at too fine-grained a level.

A few good comments solve this. Good documentation + good abstraction increases comprehensibility.